contrib/mpi-proxy-split/mpi-wrappers/mpi_collective_p2p.c

```diff
@@ -0,0 +1,638 @@
+    if (!inplace) {
+      memcpy(recvbuf + rank*recvextent*recvcount, sendbuf, sendextent*sendcount);
+    } // NOTE: if inplace, MPI guarantees that the root data is already correct
+    for (i = 0; i < size; i++) {
+      if (i != root) {
+        MPI_Recv(recvbuf + i*recvcount*recvextent, recvcount, recvtype,
+                 i, 0, comm, MPI_STATUS_IGNORE);
+      }
+    }
+  }
+  return MPI_SUCCESS;
+}
+
+int MPI_Gatherv(const void* sendbuf, int sendcount, MPI_Datatype sendtype,
+                void* recvbuf, const int recvcounts[], const int displs[],
+                MPI_Datatype recvtype, int root, MPI_Comm comm) {
+  PROLOG_rank_size;
+  int i;
+  int inplace = (sendbuf == MPI_IN_PLACE);
+  if (rank != root) {
+    for (i = 0; i < size; i++) {
+      if (i != root) {
+        MPI_Send(sendbuf, sendcount, sendtype, i, 0, comm);
+      }
+    }
+  } else { // else: rank == root
+    MPI_Aint lower_bound;
+    MPI_Aint sendextent, recvextent;
+    MPI_Type_get_extent(recvtype, &lower_bound, &recvextent);
+    if (inplace) {
+      sendextent = recvextent;
+    } else {
+      MPI_Type_get_extent(sendtype, &lower_bound, &sendextent);
+    }
+    if (!inplace) {
+      // NOTE: if inplace, MPI guarantees that the root data is already correct
+      memcpy(recvbuf + displs[root]*recvextent, sendbuf, sendextent*sendcount);
+    }
+    for (i = 0; i < size; i++) {
+      if (i != root) {
+        MPI_Recv(recvbuf + displs[i]*recvextent /* displs[i] == i*recvcount */,
+                 recvcounts[i], recvtype, i, 0, comm, MPI_STATUS_IGNORE);
+      }
+    }
+  }
+  return MPI_SUCCESS;
+}
+
+// MPI standard 3.1:  Section 5.6
+int MPI_Scatter(const void* sendbuf, int sendcount, MPI_Datatype sendtype,
+                void* recvbuf, int recvcount, MPI_Datatype recvtype,
+                int root, MPI_Comm comm) {
+  PROLOG_rank_size;
+  int i;
+  int inplace = (recvbuf == MPI_IN_PLACE);
+  if (inplace) { // if true, MPI says to ignore recvcount/recvtype
+    recvcount = sendcount;
+    recvtype = sendtype;
+  }
+  if (rank == root) {
+    MPI_Aint lower_bound;
+    MPI_Aint sendextent, recvextent;
+    MPI_Type_get_extent(sendtype, &lower_bound, &sendextent);
+    if (inplace) {
+      recvextent = sendextent;
+    } else {
+      MPI_Type_get_extent(recvtype, &lower_bound, &recvextent);
+    }
+    assert(sendextent*sendcount == recvextent*recvcount);
+    if (!inplace) {
+      memcpy(recvbuf,
+             sendbuf + rank*sendextent*sendcount,
+             sendextent*sendcount);
+    } // NOTE: if inplace, MPI guarantees that the root data is already correct
+    for (i = 0; i < size; i++) {
+      if (i != root) {
+        MPI_Send(sendbuf + i*sendcount*sendextent, sendcount, sendtype,
+                 i, 0, comm);
+      }
+    }
+  } else { // else: rank != root
+    for (i = 0; i < size; i++) {
+      if (i != root) {
+        MPI_Recv(recvbuf, recvcount, recvtype, i, 0, comm, MPI_STATUS_IGNORE);
```

Comment:
The source should be "root". Replace "I" with "root".

Reply:
@dahongli , Great catch!  Thanks.  I've pushed in a fixed version now.